c10/xpu/XPUCachingAllocator.cpp

```diff
@@ -0,0 +1,573 @@
+        stream_uses(),
+        size(size),
+        requested_size(0) {}
+
+  bool is_split() const {
+    return (prev != nullptr) || (next != nullptr);
+  }
+};
+
+bool BlockComparatorSize(const Block* a, const Block* b) {
+  if (a->queue != b->queue) {
+    return reinterpret_cast<uintptr_t>(a->queue) <
+        reinterpret_cast<uintptr_t>(b->queue);
+  }
+  if (a->size != b->size) {
+    return a->size < b->size;
+  }
+  return reinterpret_cast<uintptr_t>(a->ptr) <
+      reinterpret_cast<uintptr_t>(b->ptr);
+}
+
+struct AllocParams {
+  AllocParams(
+      DeviceIndex device,
+      size_t size,
+      sycl::queue* queue,
+      BlockPool* pool,
+      size_t alloc_size)
+      : search_key(device, queue, size),
+        pool(pool),
+        alloc_size(alloc_size),
+        block(nullptr) {}
+
+  DeviceIndex device() const {
+    return search_key.device;
+  }
+
+  sycl::queue* queue() const {
+    return search_key.queue;
+  }
+
+  size_t size() const {
+    return search_key.size;
+  }
+
+  Block search_key;
+  BlockPool* pool;
+  size_t alloc_size;
+  Block* block;
+};
+
+} // anonymous namespace
+
+class DeviceCachingAllocator {
+ private:
+  mutable std::recursive_mutex mutex;
+  BlockPool large_blocks; // unallocated cached blocks larger than 1 MB
+  BlockPool small_blocks; // unallocated cached blocks 1 MB or smaller
+  ska::flat_hash_set<Block*> active_blocks; // allocated or in use by a stream
+  ska::flat_hash_map<xpu::XPUStream, std::deque<std::pair<sycl::event, Block*>>>
+      xpu_events;
+  DeviceIndex device_index;
+
+ public:
+  DeviceCachingAllocator(DeviceIndex device_index)
+      : large_blocks(/* small */ false),
+        small_blocks(/* small */ true),
+        device_index(device_index) {}
+
+  Block* malloc(DeviceIndex device, size_t orig_size, sycl::queue& queue) {
+    std::scoped_lock<std::recursive_mutex> lock(mutex);
+    process_events();
+    size_t size = round_size(orig_size);
+    auto& pool = get_pool(size);
+    const size_t alloc_size = get_allocation_size(size);
+    AllocParams params(device, size, &queue, &pool, alloc_size);
+
+    // First, try to get a block from the existing pool.
+    bool block_found = get_free_block(params);
+    // Can't reuse an existing block, try to get a new one.
+    if (!block_found) {
+      block_found = alloc_block(params) ||
+          (release_cached_blocks() && alloc_block(params));
+    }
+    TORCH_CHECK(
+        block_found,
+        "XPU out of memory, please use `empty_cache` to release all unoccupied cached memory.");
+    bool split_remainder = should_split(params.block, params.size());
+    return alloc_found_block(std::move(params), orig_size, split_remainder);
+  }
+
+  void free(Block* block) {
+    std::scoped_lock<std::recursive_mutex> lock(mutex);
+    block->allocated = false;
+
+    if (!block->stream_uses.empty()) {
+      insert_events(block);
+    } else {
+      free_block(block);
+    }
+  }
+
+  void free_block(Block* block) {
```

Comment:
I suppose this member function should always be not invoked out of this class. Otherwise, we need to add mutex here. So we should mark this function as `private`/`protected`.

Reply:
Done. move to private scope.